refactor(api): use hl id directly in nvim_buf_set_extmark

glepnir opened the pull request:
make  `*_hl_group`  only support `nil|int|string` and convert to hl id directly.

glepnir committed:
refactor(api): use hl id directly in nvim_buf_set_extmark

bfredl commented on src/nvim/api/extmark.c:
```diff
@@ -540,36 +541,29 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer line, Integer
-    { NULL, NULL, NULL },
-  };
+  if (opts->sign_hl_group) {
+    sign.hl_id = (int)opts->sign_hl_group;
```
don't need these assignments be in if-statements. then we could just use `opts->sign_hl_group || opts->cursorline_hl_group || ...` in the end for the `sign.flags` check.

bfredl commented on src/nvim/lua/converter.c:
```diff
@@ -1324,7 +1325,27 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
+          }
+          api_free_string(name);
+          *(Integer *)mem = hl_id;
+        } else if (lua_type(L, -1) == LUA_TNUMBER) {
+          *(Integer *)mem = nlua_pop_Integer(L, err);
```
we can just do this in an `else`, after we checked for String, `nlua_pop_Integer` will do the rest.

bfredl commented on src/nvim/lua/converter.c:
```diff
@@ -1324,7 +1325,24 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
-      *(Integer *)mem = nlua_pop_Integer(L, err);
+      if (field->is_hlgroup) {
+        if (lua_type(L, -1) == LUA_TSTRING) {
+          String name = nlua_pop_String(L, err);
```
better to use `lua_tolstring` directly and avoid the alloc/free.

bfredl commented on src/nvim/lua/converter.c:
```diff
@@ -1324,7 +1325,24 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
+          }
+          api_free_string(name);
+          *(Integer *)mem = hl_id;
+        } else if (lua_type(L, -1) == LUA_TNIL) {
+          lua_pop(L, 1);
+          *(Integer *)mem = 0;
```
dead code. we don't enter the `lua_next()` loop at all for `nil` values.

bfredl commented on src/nvim/api/private/helpers.c:
```diff
@@ -936,12 +936,33 @@ bool api_dict_to_keydict(void *retval, FieldHashfn hashy, Dictionary dict, Error
+          return false;
+        }
+
+        if (value->type == kObjectTypeNil) {
```
after checking for nil, we can use `object_to_hl_id()` for both string, int handling, and also type error checking (`what` should be `k.data` )

bfredl committed:
refactor(api): give "hl_group" more accurate _meta type

These can either be number or string in lua, so we can specify this
directly as "number|string".

bfredl commented:
Nice! I pushed a small change on top to give more accurate type in `_meta/api_keysets.lua` , where we want these to be `number|string` .

glepnir commented:
perfect change!
